jetty-server/src/main/java/org/eclipse/jetty/server/Response.java

```diff
@@ -754,26 +778,57 @@ public void setStatusWithReason(int sc, String message)
     @Override
     public String getCharacterEncoding()
     {
-        if (_characterEncoding == null)
+        return getCharacterEncoding(false);
+    }
+
+    /**
+     * Private utility method to get the character encoding.
+     * A standard call to {@link #getCharacterEncoding()} should not change the Content-Type header.
+     * But when {@link #getWriter()} is called we must decide what Content Type to use, so this will allow an inferred
+     * charset to be set in in the Content-Type.
+     * @param setContentType set the Content-Type header if this is set to true, otherwise just calculate what it should be.
+     * @return the character encoding for this response.
+     */
+    private String getCharacterEncoding(boolean setContentType)
+    {
+        // First try explicit char encoding.
+        if (_characterEncoding != null)
+            return _characterEncoding;
+
+        String encoding;
+
+        // Try charset from mime type. TODO: should this be added to Content-Type header?
```

Comment:
No, because content-type has already been set, hence we have a mimeType.   We don't need to add the charset as Assumed charsets are never added to the content-type - by definition of what assumed is (otherwise they are inferred).